src/traces/histogram/calc.js

```diff
@@ -27,60 +28,35 @@ module.exports = function calc(gd, trace) {
 
     // depending on orientation, set position and size axes and data ranges
     // note: this logic for choosing orientation is duplicated in graph_obj->setstyles
-    var pos = [],
-        size = [],
-        i,
-        pa = Axes.getFromId(gd,
-            trace.orientation === 'h' ? (trace.yaxis || 'y') : (trace.xaxis || 'x')),
-        maindata = trace.orientation === 'h' ? 'y' : 'x',
-        counterdata = {x: 'y', y: 'x'}[maindata],
-        calendar = trace[maindata + 'calendar'],
-        cumulativeSpec = trace.cumulative;
+    var pos = [];
+    var size = [];
+    var pa = Axes.getFromId(gd, trace.orientation === 'h' ?
+        (trace.yaxis || 'y') : (trace.xaxis || 'x'));
+    var maindata = trace.orientation === 'h' ? 'y' : 'x';
+    var counterdata = {x: 'y', y: 'x'}[maindata];
+    var calendar = trace[maindata + 'calendar'];
+    var cumulativeSpec = trace.cumulative;
+    var i;
 
     cleanBins(trace, pa, maindata);
 
-    // prepare the raw data
-    var pos0 = pa.makeCalcdata(trace, maindata);
+    var binspec = calcAllAutoBins(gd, trace, pa, maindata);
 
-    // calculate the bins
-    var binAttr = maindata + 'bins';
-    var autoBinAttr = 'autobin' + maindata;
-    var binspec = trace[binAttr];
-    if((trace[autoBinAttr] !== false) || !binspec ||
-            binspec.start === null || binspec.end === null) {
-        binspec = Axes.autoBin(pos0, pa, trace['nbins' + maindata], false, calendar);
-
-        // adjust for CDF edge cases
-        if(cumulativeSpec.enabled && (cumulativeSpec.currentbin !== 'include')) {
-            if(cumulativeSpec.direction === 'decreasing') {
-                binspec.start = pa.c2r(pa.r2c(binspec.start) - binspec.size);
-            }
-            else {
-                binspec.end = pa.c2r(pa.r2c(binspec.end) + binspec.size);
-            }
-        }
+    // the raw data was prepared in calcAllAutoBins (during the first trace in
+    // this group) and stashed. Pull it out and drop the stash
+    var pos0 = trace._pos0;
+    delete trace._pos0;
```

Comment:
perhaps... we'd still need to stash `_pos0` because it could be used in a totally separate call to `calc`, but at least that could all be dealt with inside `calcAllAutoBins`, so yeah that would be easier to understand. 👍  (unless we end up moving to `setPositions`...)